<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 5.0.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/hexo/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/hexo/images/favicon-32x32-next.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/hexo/images/favicon-16x16-next.png">
  <link rel="mask-icon" href="/hexo/images/logo.svg" color="#222">
  <meta name="google-site-verification" content="UgODgWVmvTsdD--vjodDR3tPUZRWM3i7JpHzbYejxUM">

<link rel="stylesheet" href="/hexo/css/main.css">


<link rel="stylesheet" href="/hexo/lib/font-awesome/css/font-awesome.min.css">


<script id="hexo-configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/hexo/',
    scheme: 'Muse',
    version: '7.5.0',
    exturl: false,
    sidebar: {"position":"left","display":"post","offset":12,"onmobile":false},
    copycode: {"enable":false,"show_result":false,"style":null},
    back2top: {"enable":true,"sidebar":false,"scrollpercent":false},
    bookmark: {"enable":false,"color":"#222","save":"auto"},
    fancybox: false,
    mediumzoom: false,
    lazyload: false,
    pangu: false,
    algolia: {
      appID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    },
    localsearch: {"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},
    path: 'search.xml',
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    translation: {
      copy_button: '复制',
      copy_success: '复制成功',
      copy_failure: '复制失败'
    },
    sidebarPadding: 40
  };
</script>

  <meta name="description" content="刚好这几天要修一个异常处理的问题，回顾以下异常相关的知识点 想要解决的问题： [ ] runtime 异常和其他的异常有什么区别[ ] 异常处理的最佳实践是什么[ ] catch 里加一个 return 会怎么样 前述Java 的一个基本规范：组织糟糕的代码不会被运行。最理想的异常捕捉点是编译器，代码运行之前。但是并不是所有的 errors 都能在编译器就被捕捉到。剩下的这些问题，我们需要在运行时">
<meta property="og:type" content="article">
<meta property="og:title" content="TIJ4 通过异常处理错误 error handing with exceptions">
<meta property="og:url" content="https://jack-zheng.github.io/hexo/2021/01/21/TIJ4-error-handing-with-exceptions/index.html">
<meta property="og:site_name" content="Jackpot">
<meta property="og:description" content="刚好这几天要修一个异常处理的问题，回顾以下异常相关的知识点 想要解决的问题： [ ] runtime 异常和其他的异常有什么区别[ ] 异常处理的最佳实践是什么[ ] catch 里加一个 return 会怎么样 前述Java 的一个基本规范：组织糟糕的代码不会被运行。最理想的异常捕捉点是编译器，代码运行之前。但是并不是所有的 errors 都能在编译器就被捕捉到。剩下的这些问题，我们需要在运行时">
<meta property="og:locale" content="zh_CN">
<meta property="article:published_time" content="2021-01-21T05:49:44.000Z">
<meta property="article:modified_time" content="2021-01-25T11:50:23.449Z">
<meta property="article:author" content="Jack Zheng">
<meta property="article:tag" content="exception">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="https://jack-zheng.github.io/hexo/2021/01/21/TIJ4-error-handing-with-exceptions/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome: false,
    isPost: true,
    isPage: false,
    isArchive: false
  };
</script>

  <title>TIJ4 通过异常处理错误 error handing with exceptions | Jackpot</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-meta">

    <div>
      <a href="/hexo/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Jackpot</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
  </div>

  <div class="site-nav-toggle">
    <div class="toggle" aria-label="切换导航栏">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>
</div>


<nav class="site-nav">
  
  <ul id="menu" class="menu">
        <li class="menu-item menu-item-tags">

    <a href="/hexo/tags/" rel="section"><i class="fa fa-fw fa-tags"></i>标签</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/hexo/categories/" rel="section"><i class="fa fa-fw fa-th"></i>分类</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/hexo/archives/" rel="section"><i class="fa fa-fw fa-archive"></i>归档</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>搜索
        </a>
      </li>
  </ul>

</nav>
  <div class="site-search">
    <div class="popup search-popup">
    <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocorrect="off" autocapitalize="none"
           placeholder="搜索..." spellcheck="false"
           type="text" id="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result"></div>

</div>
<div class="search-pop-overlay"></div>

  </div>
</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>

  <a href="https://github.com/jack-zheng" class="github-corner" title="Follow me on GitHub" aria-label="Follow me on GitHub" rel="noopener" target="_blank"><svg width="80" height="80" viewBox="0 0 250 250" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path></svg></a>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content">
            

  <div class="posts-expand">
      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block " lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="https://jack-zheng.github.io/hexo/hexo/2021/01/21/TIJ4-error-handing-with-exceptions/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/hexo/images/jack.jpg">
      <meta itemprop="name" content="Jack Zheng">
      <meta itemprop="description" content="我是张江搬砖小码农, 知识搬运工">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Jackpot">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          TIJ4 通过异常处理错误 error handing with exceptions
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>

              <time title="创建时间：2021-01-21 13:49:44" itemprop="dateCreated datePublished" datetime="2021-01-21T13:49:44+08:00">2021-01-21</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                <span class="post-meta-item-text">更新于</span>
                <time title="修改时间：2021-01-25 19:50:23" itemprop="dateModified" datetime="2021-01-25T19:50:23+08:00">2021-01-25</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              <span class="post-meta-item-text">分类于</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/hexo/categories/TIJ4/" itemprop="url" rel="index">
                    <span itemprop="name">TIJ4</span>
                  </a>
                </span>
            </span>

          
            <span class="post-meta-item" title="阅读次数" id="busuanzi_container_page_pv" style="display: none;">
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              <span class="post-meta-item-text">阅读次数：</span>
              <span id="busuanzi_value_page_pv"></span>
            </span><br>
            <span class="post-meta-item" title="本文字数">
              <span class="post-meta-item-icon">
                <i class="fa fa-file-word-o"></i>
              </span>
                <span class="post-meta-item-text">本文字数：</span>
              <span>24k</span>
            </span>
            <span class="post-meta-item" title="阅读时长">
              <span class="post-meta-item-icon">
                <i class="fa fa-clock-o"></i>
              </span>
                <span class="post-meta-item-text">阅读时长 &asymp;</span>
              <span>21 分钟</span>
            </span>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <p>刚好这几天要修一个异常处理的问题，回顾以下异常相关的知识点</p>
<p>想要解决的问题：</p>
<p>[ ] runtime 异常和其他的异常有什么区别<br>[ ] 异常处理的最佳实践是什么<br>[ ] catch 里加一个 return 会怎么样</p>
<h2 id="前述"><a href="#前述" class="headerlink" title="前述"></a>前述</h2><p>Java 的一个基本规范：组织糟糕的代码不会被运行。最理想的异常捕捉点是编译器，代码运行之前。但是并不是所有的 errors 都能在编译器就被捕捉到。剩下的这些问题，我们需要在运行时让异常产生者将错误信息提交给接受者做适当的处理。</p>
<p>提升异常回复你是增加代码健壮性的重要方式。异常恢复是每个码农都需要考虑的问题，对于 Java 这种旨在为他人提供接口的语言来说，这个就尤为重要了。通过提供 error-reporting 这个错误处理模式，Java 允许组建代码将异常传给客户端代码处理。</p>
<p>Java 中的异常处理机制旨在以尽量少的代码完成经可能多的功能，并且同时让你的程序能覆盖尽可能多的异常情况。异常机制不难学，并且学会了他能马上对你的项目产生益处，因为他是唯一官方指定的处理异常的方式，并且是编译器强制检查的。</p>
<p>本章将介绍一些必须要用到异常的代码以及遇到异常时如何处理</p>
<h2 id="Concepts"><a href="#Concepts" class="headerlink" title="Concepts"></a>Concepts</h2><p>C 和其他一些早期语言经常会有多中处理异常的方式，这些方式基本都是便宜形式，并且不是语言的一部分。典型的案例就是放回一个特殊的值或者设置一个特殊的 flag，接收方根据这个返回值来进行相应的处理。但是渐渐的码农们突然意识到，自己定义的异常情况可能是不充分的，并且为了覆盖这些没有覆盖到的情况，代码会变得越来越难以维护。</p>
<p>解决方案是将异常处理正规化，这种思路的出现是一个很长过的过程，可以追溯到 1960 年。</p>
<p>exception 表示：我处理这个异常是为了。。。当异常发生时，你可能并不清楚你需要做什么，但是你知道，你不能在正常的执行下去了，你应该做点什么。你就可以将这个异常提交给上层处理，那里可能具备足够的知识处理它。</p>
<p>异常处理机制的另一个明显的好处是，它可以简化你的异常处理代码。如果没有他你必须在多个地方编写检测代码。有了它之后，就不必这么做了，exception 会保证，有人在合适的地方做检查。你只需要在一个地方处理即可。这种机制简化了你的代码，代码被分为两个分之，正常的分支和异常分支，同时你阅读，书写，调试代码也会变的更方便。</p>
<h2 id="Basic-exceptions"><a href="#Basic-exceptions" class="headerlink" title="Basic exceptions"></a>Basic exceptions</h2><p>发生异常的条件下，程序会停止执行现有的方法。区分正常程序和异常程序是很重要的，当异常产生时，你的程序由于缺少一些信息，已经不能继续执行程序了，他能做的就是跳出当前的上下文并且将执行权交给上层。</p>
<p>除 0 就是这么一种情况，当代码中出现了除 0 的情况，就需要检查一下了。可能你知道为什么要除 0，可能是你业务逻辑的需要，你知道在这种情况下需要做什么。但是如果这是一个意外情况，你必须停止当前方法并且抛出一个异常。</p>
<p>当你跑出一个异常时，有几件事情会发生。首先和其他对象一样，一个异常对象会在堆上通过 new 的方式被创建出来。然后当前的执行路径被阻断，异常对象被当前上下问弹出。exception-handling 处理机制开始接收这个异常，并试图妥善的处理它。exception handler 就是处理异常的地方，他会判断是否继续执行，或者寻求其他解决路径。</p>
<p>想象一个简单的场景，比如你有一个对象引用叫做 <code>t</code>，他可能没有被初始化过，所以你想在使用前检查一下。你可以将检查的错误通过一个对象包裹起来并 thorwing 出去，这种做法就叫做跑出异常，代码表示如下：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span>(t == <span class="keyword">null</span>)</span><br><span class="line"> <span class="keyword">throw</span> <span class="keyword">new</span> NullPointerException();</span><br></pre></td></tr></table></figure>

<p>这种做法可以让你为未来做打算，他会在之后的什么地方被处理，你很快能看到。</p>
<p>Exceptions 让你能够以 transaction 为单位处理问题， 你也可以将它想象成一个 undo 系统，你可以设置多个恢复点，当你的程序抛出异常时，他可以将程序恢复到某个稳定的节点。</p>
<p>Exceptions 最重要的一个点就是，当异常发生时，他阻止程序继续执行下去。这种情况在 C 总尤为明显，C 是没有打断机制的，这种情况下，程序可能进入到一个完全错误的状态。</p>
<h2 id="Exception-arguments"><a href="#Exception-arguments" class="headerlink" title="Exception arguments"></a>Exception arguments</h2><p>和其他 Java 中的对象一样，你可以通过 new 在堆上创建一个 exception 对象，它有两种构造函数，一种是无参的，另一种是带字符串的，比如：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">throw</span> <span class="keyword">new</span> NullPointerException(<span class="string">&quot;t = null&quot;</span>);</span><br></pre></td></tr></table></figure>

<p>当然这个信息字符串也可以在之后通过调用方法设置，之后你会看到这种用法。</p>
<p><code>throw</code> 这个关键词可以产生几种很有趣的结果。当你使用 new 创建一个 exception 的时候，你指定了 throw 的对象。虽然这个对象和你方法的返回值类型不一样，但是它还是会被这个方法返回。由此，你可以将异常处理看作一种特殊的 return 机制。返回的同时，方法和 scope 将会推出(出栈)。</p>
<p>和普通方法的共同点到此为止了，接下来的处理方式将迥异于普通方法。异常将会在 exception handler 中被处理。</p>
<p>虽然你可以在处理异常时抛出任何 Throwable 的子类，但是一般来说，我本会更具 error 的类型来指定它。error 的信息可以从他的名字和内容体现出来，但是更常见的情况是异常只包含名字而没有其他什么内容。</p>
<h2 id="Catching-an-exception"><a href="#Catching-an-exception" class="headerlink" title="Catching an exception"></a>Catching an exception</h2><p>在理解异常捕捉之前，你先得理解<strong>守护区域</strong>的概念。它代表了一段可能抛出异常的代码段，这段代码之后会紧接着一段异常处理代码。</p>
<h3 id="The-try-block"><a href="#The-try-block" class="headerlink" title="The try block"></a>The try block</h3><p>如果你在方法体中抛出一个异常(或者方法体中调用的其他方法抛出异常)，那么这个方法体在执行完 throwing 之后就结束了。如果你不想就这个结束，你可以在这些代码外面加一个 try block</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line"> <span class="comment">// Code that might generate exceptions</span></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>在不提供异常处理机制的语言中，如果你写代码很仔细的话，你可能需要为每一个方法添加异常处理，但是通过 try block 你只需要将他们全部包裹起来即可。这样你的代码会更容易阅读。</p>
<h3 id="Exception-handlers"><a href="#Exception-handlers" class="headerlink" title="Exception handlers"></a>Exception handlers</h3><p>当然，被抛出的异常都需要有一个地方来处理，这个地方就是 exception handler。它紧跟着 try block 通过关键字 catch 引出</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line"> <span class="comment">// Code that might generate exceptions</span></span><br><span class="line">&#125; <span class="keyword">catch</span>(Type1 id1)|&#123;</span><br><span class="line"> <span class="comment">// Handle exceptions of Type1</span></span><br><span class="line">&#125; <span class="keyword">catch</span>(Type2 id2) &#123;</span><br><span class="line"> <span class="comment">// Handle exceptions of Type2</span></span><br><span class="line">&#125; <span class="keyword">catch</span>(Type3 id3) &#123;</span><br><span class="line"> <span class="comment">// Handle exceptions of Type3</span></span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">// etc...</span></span><br></pre></td></tr></table></figure>

<p>每一个 catch 就是一个小的方法体，只接收一个参数。有时你甚至不需要这个参数，仅仅根据异常的名字就可以写完处理逻辑。</p>
<p>如果异常被抛出，exception-handling 机制会搜寻第一个匹配的 catch 分支并进入，当 catch 分支走完后，异常处理被视为结束。不像 switch，catch 分支不需要 break 关键字，执行完直接返回。</p>
<h2 id="Termination-vs-resumption-中断还是继续？"><a href="#Termination-vs-resumption-中断还是继续？" class="headerlink" title="Termination vs. resumption (中断还是继续？)"></a>Termination vs. resumption (中断还是继续？)</h2><p>异常处理有两种模型，Java 采用的是中断，他认为当异常发生后，你不能在回到异常发生的节点。</p>
<p>另一种是 resumption(继续)，他表示异常发生后，我们可以做一些补救措施，并且尝试重新执行失败的方法。采用这个方式意味着你在异常产生后依旧希望继续执行程序。</p>
<p>如果你想要 resumption 的处理方法，你不能在 error 发生的地方抛异常，或者你可以把你抛异常的代码放到一个循环中，多次运行，知道结果符合你预期。</p>
<p>历史上，码农们有尝试过使用 resumption 机制的操作系统，但最终回归到了 termination 机制。虽然 resumption 机制乍一听上去很美，但是并不是这么实用。可能是应为这种机制下你写的代码不能很通用，导致维护困难，特别是在写一些大型项目的时候。</p>
<h2 id="Creating-your-own-exceptions"><a href="#Creating-your-own-exceptions" class="headerlink" title="Creating your own exceptions"></a>Creating your own exceptions</h2><p>Java 允许你自己定制异常，你需要做的只是继承一个已有的异常类即可。当然继承的时候如果有的话，选一个最贴近你异常类的，那是极好的。创建时只需要用它的默认构造函数即可，代码很简单：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">SimpleException</span> <span class="keyword">extends</span> <span class="title">Exception</span> </span>&#123;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">InheritingExceptions</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">void</span> <span class="title">f</span><span class="params">()</span> <span class="keyword">throws</span> SimpleException </span>&#123;</span><br><span class="line">        System.out.println(<span class="string">&quot;Throw SimpleException from f()&quot;</span>);</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> SimpleException();</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        InheritingExceptions sed = <span class="keyword">new</span> InheritingExceptions();</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            sed.f();</span><br><span class="line">        &#125; <span class="keyword">catch</span> (SimpleException e) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;Caught it!&quot;</span>);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">// output:</span></span><br><span class="line"><span class="comment">// Throw SimpleException from f()</span></span><br><span class="line"><span class="comment">// Caught it!</span></span><br></pre></td></tr></table></figure>

<p>用以上的方式创建的异常会自带默认的构造函数，这个默认的构造函数是无参的，在自定义异常时，最重要的是要有个贴切的名字。</p>
<p>下面是带带参构造的例子</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br></pre></td><td class="code"><pre><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">MyException</span> <span class="keyword">extends</span> <span class="title">Exception</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="title">MyException</span><span class="params">()</span> </span>&#123;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="title">MyException</span><span class="params">(String msg)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">super</span>(msg);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">FullConstructors</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">f</span><span class="params">()</span> <span class="keyword">throws</span> MyException </span>&#123;</span><br><span class="line">        System.out.println(<span class="string">&quot;Throwing MyException from f()&quot;</span>);</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> MyException();</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">g</span><span class="params">()</span> <span class="keyword">throws</span> MyException </span>&#123;</span><br><span class="line">        System.out.println(<span class="string">&quot;Throwing MyException from g()&quot;</span>);</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> MyException(<span class="string">&quot;Originated in g()&quot;</span>);</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            f();</span><br><span class="line">        &#125; <span class="keyword">catch</span> (MyException e) &#123;</span><br><span class="line">            e.printStackTrace(System.out);</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            g();</span><br><span class="line">        &#125; <span class="keyword">catch</span> (MyException e) &#123;</span><br><span class="line">            e.printStackTrace(System.out);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">// output</span></span><br><span class="line"><span class="comment">// Throwing MyException from f()</span></span><br><span class="line"><span class="comment">// reading.container.MyException</span></span><br><span class="line"><span class="comment">//     at reading.container.FullConstructors.f(FullConstructors.java:15)</span></span><br><span class="line"><span class="comment">//     at reading.container.FullConstructors.main(FullConstructors.java:25)</span></span><br><span class="line"><span class="comment">// Throwing MyException from g()</span></span><br><span class="line"><span class="comment">// reading.container.MyException: Originated in g()</span></span><br><span class="line"><span class="comment">//     at reading.container.FullConstructors.g(FullConstructors.java:20)</span></span><br><span class="line"><span class="comment">//     at reading.container.FullConstructors.main(FullConstructors.java:30)</span></span><br></pre></td></tr></table></figure>

<p>只需要少量的新加 code 就能实现带参构造，声明的时候用上 <code>super</code> 关键字即可。</p>
<p>在 exception handler 中你可以看到一个方法调用叫做 <code>e.printStackTrace(System.out)</code>。他可以将异常信息输出。</p>
<h2 id="Exercises"><a href="#Exercises" class="headerlink" title="Exercises"></a>Exercises</h2><p>Exercise 1: (2) Create a class with a main( ) that throws an object of class Exception<br>inside a try block. Give the constructor for Exception a String argument. Catch the<br>exception inside a catch clause and print the String argument. Add a finally clause and<br>print a message to prove you were there.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">Exe1Exception</span> <span class="keyword">extends</span> <span class="title">Exception</span> </span>&#123;</span><br><span class="line">    Exe1Exception(String msg) &#123;</span><br><span class="line">        <span class="keyword">super</span>(msg);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">Exercise1</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            <span class="keyword">throw</span> <span class="keyword">new</span> Exe1Exception(<span class="string">&quot;my exception msg...&quot;</span>);</span><br><span class="line">        &#125; <span class="keyword">catch</span> (Exe1Exception e) &#123;</span><br><span class="line">            System.out.println(e.getMessage());</span><br><span class="line">        &#125; <span class="keyword">finally</span> &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;Into final cluster...&quot;</span>);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment">// output</span></span><br><span class="line"><span class="comment">// my exception msg...</span></span><br><span class="line"><span class="comment">// Into final cluster...</span></span><br></pre></td></tr></table></figure>

<p>Exercise 2: (1) Define an object reference and initialize it to null. Try to call a method<br>through this reference. Now wrap the code in a try-catch clause to catch the exception.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">Exercise1</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            String str = <span class="keyword">null</span>;</span><br><span class="line">            System.out.println(str.isEmpty());</span><br><span class="line">        &#125; <span class="keyword">catch</span> (NullPointerException e) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;Invoked object is null...&quot;</span>);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">// output</span></span><br><span class="line"><span class="comment">// Invoked object is null...</span></span><br></pre></td></tr></table></figure>

<p>Exercise 3: (1) Write code to generate and catch an<br>ArraylndexOutOfBoundsException.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">Exercise1</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            String[] strArr = <span class="keyword">new</span> String[<span class="number">0</span>];</span><br><span class="line">            strArr[<span class="number">0</span>] = <span class="string">&quot;str&quot;</span>;</span><br><span class="line">        &#125; <span class="keyword">catch</span> (ArrayIndexOutOfBoundsException e) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;ArrayIndexOutOfBoundsException caught...&quot;</span>);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>Exercise 4: (2) Create your own exception class using the extends keyword. Write a<br>constructor for this class that takes a String argument and stores it inside the object with a<br>String reference. Write a method that displays the stored String. Create a try-catch clause<br>to exercise your new exception.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">ExeException</span> <span class="keyword">extends</span> <span class="title">Exception</span> </span>&#123;</span><br><span class="line">    ExeException(String msg) &#123;</span><br><span class="line">        <span class="keyword">super</span>(msg);</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> String <span class="title">getExeExceptionMsg</span><span class="params">()</span> </span>&#123;</span><br><span class="line">        <span class="keyword">return</span> getMessage();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">Exercise1</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            <span class="keyword">throw</span> <span class="keyword">new</span> ExeException(<span class="string">&quot;exe exception...&quot;</span>);</span><br><span class="line">        &#125; <span class="keyword">catch</span> (ExeException e) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;Caught ExeException, msg: &quot;</span> + e.getExeExceptionMsg());</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">// output</span></span><br><span class="line"><span class="comment">// Caught ExeException, msg: exe exception...</span></span><br></pre></td></tr></table></figure>

<p>Exercise 5: (3) Create your own resumption-like behavior using a while loop that<br>repeats until an exception is no longer thrown. </p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">ExeException</span> <span class="keyword">extends</span> <span class="title">Exception</span> </span>&#123;&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">Exercise1</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">int</span> index = <span class="number">0</span>;</span><br><span class="line">        <span class="keyword">while</span> (index &lt; <span class="number">3</span>) &#123;</span><br><span class="line">            <span class="keyword">try</span> &#123;</span><br><span class="line">                System.out.println(index / <span class="number">0</span>);</span><br><span class="line">            &#125; <span class="keyword">catch</span> (ArithmeticException e) &#123;</span><br><span class="line">                System.out.println(<span class="string">&quot;Arithmetic exception when index is &quot;</span> + index);</span><br><span class="line">                index ++;</span><br><span class="line">            &#125;</span><br><span class="line">        &#125;</span><br><span class="line">        System.out.println(<span class="string">&quot;end program...&quot;</span>);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment">// output</span></span><br><span class="line"><span class="comment">// Arithmetic exception when index is 0</span></span><br><span class="line"><span class="comment">// Arithmetic exception when index is 1</span></span><br><span class="line"><span class="comment">// Arithmetic exception when index is 2</span></span><br><span class="line"><span class="comment">// end program...</span></span><br></pre></td></tr></table></figure>

<h2 id="Exceptions-and-logging"><a href="#Exceptions-and-logging" class="headerlink" title="Exceptions and logging"></a>Exceptions and logging</h2><p>使用 logging 工具类记录信息</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> java.util.logging.*;</span><br><span class="line"><span class="keyword">import</span> java.io.*;</span><br><span class="line"></span><br><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">LoggingException</span> <span class="keyword">extends</span> <span class="title">Exception</span> </span>&#123;</span><br><span class="line">    <span class="keyword">private</span> <span class="keyword">static</span> Logger logger = Logger.getLogger(<span class="string">&quot;LoggingException&quot;</span>);</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="title">LoggingException</span><span class="params">()</span> </span>&#123;</span><br><span class="line">        StringWriter trace = <span class="keyword">new</span> StringWriter();</span><br><span class="line">        printStackTrace(<span class="keyword">new</span> PrintWriter(trace));</span><br><span class="line">        logger.severe(trace.toString());</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">LoggingExceptions</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            <span class="keyword">throw</span> <span class="keyword">new</span> LoggingException();</span><br><span class="line">        &#125; <span class="keyword">catch</span> (LoggingException e) &#123;</span><br><span class="line">            System.err.println(<span class="string">&quot;Caught &quot;</span> + e);</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            <span class="keyword">throw</span> <span class="keyword">new</span> LoggingException();</span><br><span class="line">        &#125; <span class="keyword">catch</span> (LoggingException e) &#123;</span><br><span class="line">            System.err.println(<span class="string">&quot;Caught &quot;</span> + e);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">// output</span></span><br><span class="line"><span class="comment">// Jan 25, 2021 10:57:56 AM reading.container.LoggingException &lt;init&gt;</span></span><br><span class="line"><span class="comment">// SEVERE: reading.container.LoggingException</span></span><br><span class="line"><span class="comment">//     at reading.container.LoggingExceptions.main(LoggingExceptions.java:20)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">// Caught reading.container.LoggingException</span></span><br><span class="line"><span class="comment">// Jan 25, 2021 10:57:56 AM reading.container.LoggingException &lt;init&gt;</span></span><br><span class="line"><span class="comment">// SEVERE: reading.container.LoggingException</span></span><br><span class="line"><span class="comment">//     at reading.container.LoggingExceptions.main(LoggingExceptions.java:25)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">// Caught reading.container.LoggingException</span></span><br></pre></td></tr></table></figure>

<p><code>Logger.getLogger()</code> 接收一个字符串作为参数创建 Logger 对象，如果没有其他设置，他回把对应的信息输出到 <code>System.err</code> 中去。<code>printStackTrace()</code> 接收一个 PrintWriter 作为参数，再通过调用 <code>logger.severe()</code> 将信息输出。</p>
<p>上面这种处理方式将所有的 logging 相关动作封装在了异常中，所以很简便，但是更常见的处理方式是将你要处理的 log 在 exception handler 中进行封装。</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br></pre></td><td class="code"><pre><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">MyException2</span> <span class="keyword">extends</span> <span class="title">Exception</span> </span>&#123;</span><br><span class="line">    <span class="keyword">private</span> <span class="keyword">int</span> x;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="title">MyException2</span><span class="params">()</span> </span>&#123;&#125;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="title">MyException2</span><span class="params">(String msg)</span> </span>&#123; <span class="keyword">super</span>(msg); &#125;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="title">MyException2</span><span class="params">(String msg, <span class="keyword">int</span> x)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">super</span>(msg);</span><br><span class="line">        <span class="keyword">this</span>.x = x;</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">int</span> <span class="title">val</span><span class="params">()</span> </span>&#123; <span class="keyword">return</span> x; &#125;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> String <span class="title">getMessage</span><span class="params">()</span> </span>&#123;</span><br><span class="line">        <span class="keyword">return</span> <span class="string">&quot;Detail Message: &quot;</span>+ x + <span class="string">&quot; &quot;</span>+ <span class="keyword">super</span>.getMessage();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">ExtraFeatures</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">f</span><span class="params">()</span> <span class="keyword">throws</span> MyException2 </span>&#123;</span><br><span class="line">        System.out.println(<span class="string">&quot;Throwing MyException2 from f()&quot;</span>);</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> MyException2();</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">g</span><span class="params">()</span> <span class="keyword">throws</span> MyException2 </span>&#123;</span><br><span class="line">        System.out.println(<span class="string">&quot;Throwing MyException2 from g()&quot;</span>);</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> MyException2(<span class="string">&quot;Originated in g()&quot;</span>);</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">h</span><span class="params">()</span> <span class="keyword">throws</span> MyException2 </span>&#123;</span><br><span class="line">        System.out.println(<span class="string">&quot;Throwing MyException2 from h()&quot;</span>);</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> MyException2(<span class="string">&quot;Originated in h()&quot;</span>, <span class="number">47</span>);</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            f();</span><br><span class="line">        &#125; <span class="keyword">catch</span>(MyException2 e) &#123;</span><br><span class="line">            e.printStackTrace(System.out);</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            g();</span><br><span class="line">        &#125; <span class="keyword">catch</span>(MyException2 e) &#123;</span><br><span class="line">            e.printStackTrace(System.out);</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            h();</span><br><span class="line">        &#125; <span class="keyword">catch</span>(MyException2 e) &#123;</span><br><span class="line">            e.printStackTrace(System.out);</span><br><span class="line">            System.out.println(<span class="string">&quot;e.val() = &quot;</span> + e.val());</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">// output</span></span><br><span class="line"><span class="comment">// Throwing MyException2 from f()</span></span><br><span class="line"><span class="comment">// reading.container.MyException2: Detail Message: 0 null</span></span><br><span class="line"><span class="comment">//     at reading.container.ExtraFeatures.f(ExtraFeatures.java:20)</span></span><br><span class="line"><span class="comment">//     at reading.container.ExtraFeatures.main(ExtraFeatures.java:32)</span></span><br><span class="line"><span class="comment">// Throwing MyException2 from g()</span></span><br><span class="line"><span class="comment">// reading.container.MyException2: Detail Message: 0 Originated in g()</span></span><br><span class="line"><span class="comment">//     at reading.container.ExtraFeatures.g(ExtraFeatures.java:24)</span></span><br><span class="line"><span class="comment">//     at reading.container.ExtraFeatures.main(ExtraFeatures.java:37)</span></span><br><span class="line"><span class="comment">// Throwing MyException2 from h()</span></span><br><span class="line"><span class="comment">// reading.container.MyException2: Detail Message: 47 Originated in h()</span></span><br><span class="line"><span class="comment">//     at reading.container.ExtraFeatures.h(ExtraFeatures.java:28)</span></span><br><span class="line"><span class="comment">//     at reading.container.ExtraFeatures.main(ExtraFeatures.java:42)</span></span><br><span class="line"><span class="comment">// e.val() = 47</span></span><br></pre></td></tr></table></figure>

<p>我们在自定义异常中添加了一个新的 field 并给出了对应的 getter 方法，重写了 <code>Throwable.getMessage( )</code> 方法。</p>
<p>exception 也是一个 Java 对象，你可以继续扩展这个类，但是值得注意的是，你的包装可能被其他人忽略，因为他们在使用的时候可能只想找一个贴切的异常并丢出去。</p>
<h2 id="The-exception-specification"><a href="#The-exception-specification" class="headerlink" title="The exception specification"></a>The exception specification</h2><p>在 Java 中，你需要告知调用者你的方法可能会抛出什么异常，而且这是强制的语法。这种语法使用 throw 作为关键字，后面接需要 catch 的异常 <code>void f() throws TooBig, TooSmall, DivZero &#123; //...</code></p>
<p>如果方法声明只是简单的 <code>void f() &#123; //...</code> 这表示没有异常从这个方法中抛出。 <code>&#123;except</code> 表示异常继承自 <code>RuntimeException</code>，这个异常可以在任何地方抛出。在异常声明中，你不能作弊。如果你方法中有抛出异常，但是你没有处理的话，编译器就会监测到并给你提示，要跑处理它，要跑继续抛出它。通过自顶向下的约束异常声明，Java 保证了在编译期的异常检测。</p>
<p>有一个特别的地方是，你可以在没有对应实现的情况下抛出异常。这种处理方式可以看作是一个预先打桩，为你将来的实现做预备，而且省去了你以后还要改应用代码的麻烦。</p>
<p>在编译期强制你检测的这种异常，叫做 Checked Exception。</p>
<h2 id="Catching-any-exception"><a href="#Catching-any-exception" class="headerlink" title="Catching any exception"></a>Catching any exception</h2><p>在异常处理中，声明一个 catch 来捕捉所有的异常是可行的。你可以 catch Exception 来实现，他是几乎所有异常的基类</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">catch</span>(Exception e) &#123;</span><br><span class="line"> System.out.println(<span class="string">&quot;Caught an exception&quot;</span>);</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>他会处理任何异常，所以确保将它放到你的 catch 列表的末位。由于他是一个基类，所以你一般不能得到什么很特殊的信息，但是你还是可以调用那些基于 Throwable 的方法，比如</p>
<p>String getMessage( )<br>String getLocalizedMessage( )  </p>
<p>获取 message，或者是基于本地化的 message。</p>
<p><code>String toString( )</code> 返回一个间断的关于 Throwable 类的描述，如果这个类有详细信息的话，也会包含在其中。</p>
<p>void printStackTrace( )<br>void printStackTrace(PrintStream)<br>void printStackTrace(java.io.PrintWriter)  </p>
<p>打印 Throwable 以及对应的调用栈信息。栈信息会告诉你异常发生的点。第一种方式会将异常输出到 standard error, 第二和三种方式会输出到对应的流。</p>
<p>Throwable 还有很多其他的方法可以调用，比如 <code>getClass()</code>， 它能返回一个异常对象，<code>getName()</code> 返回类信息，包含路径名，<code>getSimpleName()</code> 只含有类名。</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">ExceptionMethods</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            <span class="keyword">throw</span> <span class="keyword">new</span> Exception(<span class="string">&quot;My Exception&quot;</span>);</span><br><span class="line">        &#125; <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;Caught Exception&quot;</span>);</span><br><span class="line">            System.out.println(<span class="string">&quot;getMessage():&quot;</span> + e.getMessage());</span><br><span class="line">            System.out.println(<span class="string">&quot;getLocalizedMessage():&quot;</span> +</span><br><span class="line">                    e.getLocalizedMessage());</span><br><span class="line">            System.out.println(<span class="string">&quot;toString():&quot;</span> + e);</span><br><span class="line">            System.out.println(<span class="string">&quot;System.out.printlnStackTrace():&quot;</span>);</span><br><span class="line">            e.printStackTrace(System.out);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">// output</span></span><br><span class="line"><span class="comment">// Caught Exception</span></span><br><span class="line"><span class="comment">// getMessage():My Exception</span></span><br><span class="line"><span class="comment">// getLocalizedMessage():My Exception</span></span><br><span class="line"><span class="comment">// toString():java.lang.Exception: My Exception</span></span><br><span class="line"><span class="comment">// System.out.printlnStackTrace():</span></span><br><span class="line"><span class="comment">// java.lang.Exception: My Exception</span></span><br><span class="line"><span class="comment">//     at reading.container.ExceptionMethods.main(ExceptionMethods.java:6)</span></span><br></pre></td></tr></table></figure>

<h3 id="The-stack-trace"><a href="#The-stack-trace" class="headerlink" title="The stack trace"></a>The stack trace</h3><p><code>printStackTrace( )</code> 中的信息也可以通过 <code>getStackTrace( )</code> 得到，他会返回一个信息栈。下面是一个示例，可以看到，root cause 是在第一行打印的，最外层的异常点在最后打印。</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">WhoCalled</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">static</span> <span class="keyword">void</span> <span class="title">f</span><span class="params">()</span> </span>&#123;</span><br><span class="line">        <span class="comment">// Generate an exception to fill in the stack trace</span></span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            <span class="keyword">throw</span> <span class="keyword">new</span> Exception();</span><br><span class="line">        &#125; <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            <span class="keyword">for</span> (StackTraceElement ste : e.getStackTrace())</span><br><span class="line">                System.out.println(ste.getMethodName());</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">static</span> <span class="keyword">void</span> <span class="title">g</span><span class="params">()</span> </span>&#123;</span><br><span class="line">        f();</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">static</span> <span class="keyword">void</span> <span class="title">h</span><span class="params">()</span> </span>&#123;</span><br><span class="line">        g();</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        f();</span><br><span class="line">        System.out.println(<span class="string">&quot;--------------------------------&quot;</span>);</span><br><span class="line">        g();</span><br><span class="line">        System.out.println(<span class="string">&quot;--------------------------------&quot;</span>);</span><br><span class="line">        h();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment">// output</span></span><br><span class="line"><span class="comment">// f</span></span><br><span class="line"><span class="comment">// main</span></span><br><span class="line"><span class="comment">// --------------------------------</span></span><br><span class="line"><span class="comment">// f</span></span><br><span class="line"><span class="comment">// g</span></span><br><span class="line"><span class="comment">// main</span></span><br><span class="line"><span class="comment">// --------------------------------</span></span><br><span class="line"><span class="comment">// f</span></span><br><span class="line"><span class="comment">// g</span></span><br><span class="line"><span class="comment">// h</span></span><br><span class="line"><span class="comment">// main</span></span><br></pre></td></tr></table></figure>

<h3 id="Rethrowing-an-exception"><a href="#Rethrowing-an-exception" class="headerlink" title="Rethrowing an exception"></a>Rethrowing an exception</h3><p>有时你在捕捉到异常之后会想要再一次 throw 它，比如之前提到的，通过 Exception 捕捉到异常的情况。这时你只需要在 handler 里面再 throw 即可</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">catch</span>(Exception e) &#123;</span><br><span class="line">    System.out.println(<span class="string">&quot;An exception was thrown&quot;</span>);</span><br><span class="line">    <span class="keyword">throw</span> e;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>Rethrowing 会将异常交由更高的 context 处理，这个过程中，异常对象的所有信息都会被保存下来，如果你想要创建一个新的异常对象，你可以使用 <code>fillInStackTrace( )</code> 方法，示例如下：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">Rethrowing</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">f</span><span class="params">()</span> <span class="keyword">throws</span> Exception </span>&#123;</span><br><span class="line">        System.out.println(<span class="string">&quot;originating the exception in f()&quot;</span>);</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> Exception(<span class="string">&quot;thrown from f()&quot;</span>);</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">g</span><span class="params">()</span> <span class="keyword">throws</span> Exception </span>&#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            f();</span><br><span class="line">        &#125; <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;Inside g(),e.printStackTrace()&quot;</span>);</span><br><span class="line">            e.printStackTrace(System.out);</span><br><span class="line">            <span class="keyword">throw</span> e;</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">h</span><span class="params">()</span> <span class="keyword">throws</span> Exception </span>&#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            f();</span><br><span class="line">        &#125; <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;Inside h(),e.printStackTrace()&quot;</span>);</span><br><span class="line">            e.printStackTrace(System.out);</span><br><span class="line">            <span class="keyword">throw</span> (Exception) e.fillInStackTrace();</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            g();</span><br><span class="line">        &#125; <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;main: printStackTrace()&quot;</span>);</span><br><span class="line">            e.printStackTrace(System.out);</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            h();</span><br><span class="line">        &#125; <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;main: printStackTrace()&quot;</span>);</span><br><span class="line">            e.printStackTrace(System.out);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment">// originating the exception in f()</span></span><br><span class="line"><span class="comment">// Inside g(),e.printStackTrace()</span></span><br><span class="line"><span class="comment">// java.lang.Exception: thrown from f()</span></span><br><span class="line"><span class="comment">//     at reading.container.Rethrowing.f(Rethrowing.java:6)</span></span><br><span class="line"><span class="comment">//     at reading.container.Rethrowing.g(Rethrowing.java:11)</span></span><br><span class="line"><span class="comment">//     at reading.container.Rethrowing.main(Rethrowing.java:31)</span></span><br><span class="line"><span class="comment">// main: printStackTrace()</span></span><br><span class="line"><span class="comment">// java.lang.Exception: thrown from f()</span></span><br><span class="line"><span class="comment">//     at reading.container.Rethrowing.f(Rethrowing.java:6)</span></span><br><span class="line"><span class="comment">//     at reading.container.Rethrowing.g(Rethrowing.java:11)</span></span><br><span class="line"><span class="comment">//     at reading.container.Rethrowing.main(Rethrowing.java:31)</span></span><br><span class="line"><span class="comment">// /--------------------- Dash --------------------/</span></span><br><span class="line"><span class="comment">// originating the exception in f()</span></span><br><span class="line"><span class="comment">// Inside h(),e.printStackTrace()</span></span><br><span class="line"><span class="comment">// java.lang.Exception: thrown from f()</span></span><br><span class="line"><span class="comment">//     at reading.container.Rethrowing.f(Rethrowing.java:6)</span></span><br><span class="line"><span class="comment">//     at reading.container.Rethrowing.h(Rethrowing.java:21)</span></span><br><span class="line"><span class="comment">//     at reading.container.Rethrowing.main(Rethrowing.java:38)</span></span><br><span class="line"><span class="comment">// main: printStackTrace()</span></span><br><span class="line"><span class="comment">// java.lang.Exception: thrown from f()</span></span><br><span class="line"><span class="comment">//     at reading.container.Rethrowing.h(Rethrowing.java:25)</span></span><br><span class="line"><span class="comment">//     at reading.container.Rethrowing.main(Rethrowing.java:38)</span></span><br></pre></td></tr></table></figure>

<p><code>f()</code> 中通过 <code>fillInStackTrace( )</code> 改变了异常原点，相比于之前的调用 <code>g()</code> 的方法信息没有了。当然你也可以用 throw 新的 Exception 来实现和 <code>fillInStackTrace()</code> 同样的功能</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br></pre></td><td class="code"><pre><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">OneException</span> <span class="keyword">extends</span> <span class="title">Exception</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="title">OneException</span><span class="params">(String s)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">super</span>(s);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">TwoException</span> <span class="keyword">extends</span> <span class="title">Exception</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="title">TwoException</span><span class="params">(String s)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">super</span>(s);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">RethrowNew</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">f</span><span class="params">()</span> <span class="keyword">throws</span> OneException </span>&#123;</span><br><span class="line">        System.out.println(<span class="string">&quot;originating the exception in f()&quot;</span>);</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> OneException(<span class="string">&quot;thrown from f()&quot;</span>);</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            <span class="keyword">try</span> &#123;</span><br><span class="line">                f();</span><br><span class="line">            &#125; <span class="keyword">catch</span> (OneException e) &#123;</span><br><span class="line">                System.out.println(</span><br><span class="line">                        <span class="string">&quot;Caught in inner try, e.printStackTrace()&quot;</span>);</span><br><span class="line">                e.printStackTrace(System.out);</span><br><span class="line">                <span class="keyword">throw</span> <span class="keyword">new</span> TwoException(<span class="string">&quot;from inner try&quot;</span>);</span><br><span class="line">            &#125;</span><br><span class="line">        &#125; <span class="keyword">catch</span> (TwoException e) &#123;</span><br><span class="line">            System.out.println(</span><br><span class="line">                    <span class="string">&quot;Caught in outer try, e.printStackTrace()&quot;</span>);</span><br><span class="line">            e.printStackTrace(System.out);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment">// originating the exception in f()</span></span><br><span class="line"><span class="comment">// Caught in inner try, e.printStackTrace()</span></span><br><span class="line"><span class="comment">// reading.container.OneException: thrown from f()</span></span><br><span class="line"><span class="comment">//     at reading.container.RethrowNew.f(RethrowNew.java:18)</span></span><br><span class="line"><span class="comment">//     at reading.container.RethrowNew.main(RethrowNew.java:24)</span></span><br><span class="line"><span class="comment">// Caught in outer try, e.printStackTrace()</span></span><br><span class="line"><span class="comment">// reading.container.TwoException: from inner try</span></span><br><span class="line"><span class="comment">//     at reading.container.RethrowNew.main(RethrowNew.java:29)</span></span><br></pre></td></tr></table></figure>

<p>最后的 exception handler 只知道异常来源于 inner try block 而不知道任何关于 f() 的信息。你完全不用关心异常的清理问题，他们都是基于堆创建的对象，垃圾回收机制会负责清理他们。</p>
<h3 id="Exception-chaining"><a href="#Exception-chaining" class="headerlink" title="Exception chaining"></a>Exception chaining</h3><p>通常来说，当你抛出自己的异常时，你都会希望这个异常带有原始异常的信息。在 Java 1.4 以前，码农们需要自己处理这个问题，但是之后的版本中，你可以通过在构造函数中传入异常类来实现这个功能。</p>
<p>Throwable 的子类中只有三个提供这个功能，分别是 Error(用于记录 JVM 异常)，Exception 和 RuntimeException。如果其他类型的异常，你也想串联起来的话，你可以调用 <code>initCause()</code> 方法，示例如下：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br></pre></td><td class="code"><pre><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">DynamicFieldsException</span> <span class="keyword">extends</span> <span class="title">Exception</span> </span>&#123;&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">DynamicFields</span> </span>&#123;</span><br><span class="line">    <span class="keyword">private</span> Object[][] fields;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="title">DynamicFields</span><span class="params">(<span class="keyword">int</span> initialSize)</span> </span>&#123;</span><br><span class="line">        fields = <span class="keyword">new</span> Object[initialSize][<span class="number">2</span>];</span><br><span class="line">        <span class="keyword">for</span> (<span class="keyword">int</span> i = <span class="number">0</span>; i &lt; initialSize; i++)</span><br><span class="line">            fields[i] = <span class="keyword">new</span> Object[]&#123;<span class="keyword">null</span>, <span class="keyword">null</span>&#125;;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> String <span class="title">toString</span><span class="params">()</span> </span>&#123;</span><br><span class="line">        StringBuilder result = <span class="keyword">new</span> StringBuilder();</span><br><span class="line">        <span class="keyword">for</span> (Object[] obj : fields) &#123;</span><br><span class="line">            result.append(obj[<span class="number">0</span>]);</span><br><span class="line">            result.append(<span class="string">&quot;: &quot;</span>);</span><br><span class="line">            result.append(obj[<span class="number">1</span>]);</span><br><span class="line">            result.append(<span class="string">&quot;\n&quot;</span>);</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">return</span> result.toString();</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">private</span> <span class="keyword">int</span> <span class="title">hasField</span><span class="params">(String id)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">for</span> (<span class="keyword">int</span> i = <span class="number">0</span>; i &lt; fields.length; i++)</span><br><span class="line">            <span class="keyword">if</span> (id.equals(fields[i][<span class="number">0</span>]))</span><br><span class="line">                <span class="keyword">return</span> i;</span><br><span class="line">        <span class="keyword">return</span> -<span class="number">1</span>;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">private</span> <span class="keyword">int</span></span><br><span class="line">    getFieldNumber(String id) <span class="keyword">throws</span> NoSuchFieldException &#123;</span><br><span class="line">        <span class="keyword">int</span> fieldNum = hasField(id);</span><br><span class="line">        <span class="keyword">if</span> (fieldNum == -<span class="number">1</span>)</span><br><span class="line">            <span class="keyword">throw</span> <span class="keyword">new</span> NoSuchFieldException();</span><br><span class="line">        <span class="keyword">return</span> fieldNum;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">private</span> <span class="keyword">int</span> <span class="title">makeField</span><span class="params">(String id)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">for</span> (<span class="keyword">int</span> i = <span class="number">0</span>; i &lt; fields.length; i++)</span><br><span class="line">            <span class="keyword">if</span> (fields[i][<span class="number">0</span>] == <span class="keyword">null</span>) &#123;</span><br><span class="line">                fields[i][<span class="number">0</span>] = id;</span><br><span class="line">                <span class="keyword">return</span> i;</span><br><span class="line">            &#125;</span><br><span class="line">        <span class="comment">// No empty fields. Add one:</span></span><br><span class="line">        Object[][] tmp = <span class="keyword">new</span> Object[fields.length + <span class="number">1</span>][<span class="number">2</span>];</span><br><span class="line">        <span class="keyword">for</span> (<span class="keyword">int</span> i = <span class="number">0</span>; i &lt; fields.length; i++)</span><br><span class="line">            tmp[i] = fields[i];</span><br><span class="line">        <span class="keyword">for</span> (<span class="keyword">int</span> i = fields.length; i &lt; tmp.length; i++)</span><br><span class="line">            tmp[i] = <span class="keyword">new</span> Object[]&#123;<span class="keyword">null</span>, <span class="keyword">null</span>&#125;;</span><br><span class="line">        fields = tmp;</span><br><span class="line">        <span class="comment">// Recursive call with expanded fields:</span></span><br><span class="line">        <span class="keyword">return</span> makeField(id);</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> Object <span class="title">getField</span><span class="params">(String id)</span> <span class="keyword">throws</span> NoSuchFieldException </span>&#123;</span><br><span class="line">        <span class="keyword">return</span> fields[getFieldNumber(id)][<span class="number">1</span>];</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> Object <span class="title">setField</span><span class="params">(String id, Object value)</span></span></span><br><span class="line"><span class="function">            <span class="keyword">throws</span> DynamicFieldsException </span>&#123;</span><br><span class="line">        <span class="keyword">if</span> (value == <span class="keyword">null</span>) &#123;</span><br><span class="line">            <span class="comment">// Most exceptions don’t have a &quot;cause&quot; constructor.</span></span><br><span class="line">            <span class="comment">// In these cases you must use initCause(),</span></span><br><span class="line">            <span class="comment">// available in all Throwable subclasses.</span></span><br><span class="line">            DynamicFieldsException dfe = <span class="keyword">new</span> DynamicFieldsException();</span><br><span class="line">            dfe.initCause(<span class="keyword">new</span> NullPointerException());</span><br><span class="line">            <span class="keyword">throw</span> dfe;</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">int</span> fieldNumber = hasField(id);</span><br><span class="line">        <span class="keyword">if</span> (fieldNumber == -<span class="number">1</span>)</span><br><span class="line">            fieldNumber = makeField(id);</span><br><span class="line">        Object result = <span class="keyword">null</span>;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            result = getField(id); <span class="comment">// Get old value</span></span><br><span class="line">        &#125; <span class="keyword">catch</span> (NoSuchFieldException e) &#123;</span><br><span class="line">            <span class="comment">// Use constructor that takes &quot;cause&quot;:</span></span><br><span class="line">            <span class="keyword">throw</span> <span class="keyword">new</span> RuntimeException(e);</span><br><span class="line">        &#125;</span><br><span class="line">        fields[fieldNumber][<span class="number">1</span>] = value;</span><br><span class="line">        <span class="keyword">return</span> result;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        DynamicFields df = <span class="keyword">new</span> DynamicFields(<span class="number">3</span>);</span><br><span class="line">        System.out.println(df);</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            df.setField(<span class="string">&quot;d&quot;</span>, <span class="string">&quot;A value for d&quot;</span>);</span><br><span class="line">            df.setField(<span class="string">&quot;number&quot;</span>, <span class="number">47</span>);</span><br><span class="line">            df.setField(<span class="string">&quot;number2&quot;</span>, <span class="number">48</span>);</span><br><span class="line">            System.out.println(df);</span><br><span class="line">            df.setField(<span class="string">&quot;d&quot;</span>, <span class="string">&quot;A new value for d&quot;</span>);</span><br><span class="line">            df.setField(<span class="string">&quot;number3&quot;</span>, <span class="number">11</span>);</span><br><span class="line">            System.out.println(<span class="string">&quot;df: &quot;</span> + df);</span><br><span class="line">            System.out.println(<span class="string">&quot;df.getField(\&quot;d\&quot;) : &quot;</span> + df.getField(<span class="string">&quot;d&quot;</span>));</span><br><span class="line">            Object field = df.setField(<span class="string">&quot;d&quot;</span>, <span class="keyword">null</span>); <span class="comment">// Exception</span></span><br><span class="line">        &#125; <span class="keyword">catch</span> (NoSuchFieldException e) &#123;</span><br><span class="line">            e.printStackTrace(System.out);</span><br><span class="line">        &#125; <span class="keyword">catch</span> (DynamicFieldsException e) &#123;</span><br><span class="line">            e.printStackTrace(System.out);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment">// output</span></span><br><span class="line"><span class="comment">// null: null</span></span><br><span class="line"><span class="comment">// null: null</span></span><br><span class="line"><span class="comment">// null: null</span></span><br><span class="line"></span><br><span class="line"><span class="comment">// d: A value for d</span></span><br><span class="line"><span class="comment">// number: 47</span></span><br><span class="line"><span class="comment">// number2: 48</span></span><br><span class="line"></span><br><span class="line"><span class="comment">// df: d: A new value for d</span></span><br><span class="line"><span class="comment">// number: 47</span></span><br><span class="line"><span class="comment">// number2: 48</span></span><br><span class="line"><span class="comment">// number3: 11</span></span><br><span class="line"></span><br><span class="line"><span class="comment">// df.getField(&quot;d&quot;) : A new value for d</span></span><br><span class="line"><span class="comment">// reading.container.DynamicFieldsException</span></span><br><span class="line"><span class="comment">//     at reading.container.DynamicFields.setField(DynamicFields.java:68)</span></span><br><span class="line"><span class="comment">//     at reading.container.DynamicFields.main(DynamicFields.java:98)</span></span><br><span class="line"><span class="comment">// Caused by: java.lang.NullPointerException</span></span><br><span class="line"><span class="comment">//     at reading.container.DynamicFields.setField(DynamicFields.java:69)</span></span><br><span class="line"><span class="comment">//     ... 1 more</span></span><br></pre></td></tr></table></figure>

<p>示例说明：</p>
<ul>
<li>自定义一个异常 DynamicFieldsException</li>
<li>DynamicFields 为测试类，包含一个需要处理的 field 叫做 fields，他是一个二维数组</li>
<li>fields 初始化时可以给定长度，宽度为固定值 2, 也就是 n*2 的矩阵</li>
<li>fields 的子单元值为对象，不能填充原始类型的值</li>
<li>自定义 toString 方法可以答应矩阵值</li>
<li>setField 可以设置一行的值，如果超出容量，自动 copy + append, 设置的值不能为 null 否则报错</li>
<li>getField 返回对应行的值，如果没有抛异常</li>
</ul>
<p>在 <code>setField()</code> 方法中，我们我们为 DynamicFieldsException 通过调用 initCause 设置了 NPE 为 root</p>
<h2 id="Standard-Java-exceptions"><a href="#Standard-Java-exceptions" class="headerlink" title="Standard Java exceptions"></a>Standard Java exceptions</h2><p>Java 的 Throwable 类代表了所有可 throw 类，通常有两个常用子类 Error 和 Exception。Error 表示 compile-time 和系统错误，这些是你不需要关心的。另一类是 Exception，这些是码农需要关心的。</p>
<p>想要对 Exception 有一个概览，最好就去看一下 JDK 文档，这可以给你找找感觉，但是当你看了之后，你会发现，这些异常，出了名字不同外，其他基本都是一样的。如果你是用第三方包，那么很大概率会遇到他们自定义的异常。所以最重要的事是来哦姐他的定义，还有就是知道当你遇到它时你需要做什么。</p>
<p>异常的名字就代表了它处理的场景，异常的命名要求贴切明了。异常并不是全都定义在 java.lang 下，其他一些包，比如 util, net 和 io 也都有自己的异常类。你可以通过查看他们的包路径知道这些信息。比如所有的 I/O 异常都是继承于 java.io.IOException。</p>
<h3 id="Special-case-RuntimeException"><a href="#Special-case-RuntimeException" class="headerlink" title="Special case: RuntimeException"></a>Special case: RuntimeException</h3><p>下面是第一个示例</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span> (t == <span class="keyword">null</span>) &#123;</span><br><span class="line">    <span class="keyword">throw</span> <span class="keyword">new</span> NullPointException();</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>如果代码中每个可能有 null 引用的地方都需要做 NPE 检测，那想象就很刺激。所幸，这个检测 Java 会替你完成，所以上述的代码中的 NPE 检查是多余的。</p>
<p>JDK 中有一族异常处理类似的问题，Java 代码中会自动抛出，自动处理这些异常签名。他们有一个基类叫做 <code>RuntimeException</code>, 由它派生出来的异常都不需要在声明中他别指出来。他们也被叫做 unchecked exceptions(非受检异常)。虽然你不需要检测 RuntimeException, 都是你在写代码的过程中可能会想要抛出这个异常。</p>
<p>下面是一个没有捕获 RuntimeException 的例子：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">NeverCaught</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">static</span> <span class="keyword">void</span> <span class="title">f</span><span class="params">()</span> </span>&#123;</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> RuntimeException(<span class="string">&quot;From f()&quot;</span>);</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">static</span> <span class="keyword">void</span> <span class="title">g</span><span class="params">()</span> </span>&#123;</span><br><span class="line">        f();</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        g();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment">// output</span></span><br><span class="line"><span class="comment">// Exception in thread &quot;main&quot; java.lang.RuntimeException: From f()</span></span><br><span class="line"><span class="comment">//     at reading.container.NeverCaught.f(NeverCaught.java:5)</span></span><br><span class="line"><span class="comment">//     at reading.container.NeverCaught.g(NeverCaught.java:9)</span></span><br><span class="line"><span class="comment">//     at reading.container.NeverCaught.main(NeverCaught.java:13)</span></span><br></pre></td></tr></table></figure>

<p>你可以看到，即使你在 f() 中 throw 了这个异常，但是你在调用它的位置也不需要用异常签名标识它。</p>
<p>时刻牢记，只有 运行时异常 可以这么处理， checked exception 不行，因为 Java 语法中，将运行时异常当作系统错误处理，系统错误的定义：</p>
<ol>
<li>那些你不能预料的异常，比如 null reference</li>
<li>那种作为作者，你在程序中应该检查的错误，比如 ArraylndexOutOfBoundsException</li>
</ol>
<h2 id="Performing-cleanup-with-finally"><a href="#Performing-cleanup-with-finally" class="headerlink" title="Performing cleanup with finally"></a>Performing cleanup with finally</h2><h3 id="What’s-finally-for"><a href="#What’s-finally-for" class="headerlink" title="What’s finally for?"></a>What’s finally for?</h3><h3 id="Using-finally-during-return"><a href="#Using-finally-during-return" class="headerlink" title="Using finally during return"></a>Using finally during return</h3><h3 id="Pitfall-the-lost-exception"><a href="#Pitfall-the-lost-exception" class="headerlink" title="Pitfall: the lost exception"></a>Pitfall: the lost exception</h3><h2 id="Exception-restrictions"><a href="#Exception-restrictions" class="headerlink" title="Exception restrictions"></a>Exception restrictions</h2><h2 id="Constructors"><a href="#Constructors" class="headerlink" title="Constructors"></a>Constructors</h2><h2 id="Exception-matching"><a href="#Exception-matching" class="headerlink" title="Exception matching"></a>Exception matching</h2><h2 id="Alternative-approaches"><a href="#Alternative-approaches" class="headerlink" title="Alternative approaches"></a>Alternative approaches</h2><h3 id="History"><a href="#History" class="headerlink" title="History"></a>History</h3><h3 id="Perspectives"><a href="#Perspectives" class="headerlink" title="Perspectives"></a>Perspectives</h3><h3 id="Passing-exceptions-to-the-console"><a href="#Passing-exceptions-to-the-console" class="headerlink" title="Passing exceptions to the console"></a>Passing exceptions to the console</h3><h3 id="Converting-checked-to-unchecked-exceptions"><a href="#Converting-checked-to-unchecked-exceptions" class="headerlink" title="Converting checked to unchecked exceptions"></a>Converting checked to unchecked exceptions</h3><h2 id="Exception-guidelines"><a href="#Exception-guidelines" class="headerlink" title="Exception guidelines"></a>Exception guidelines</h2><h2 id="Summary"><a href="#Summary" class="headerlink" title="Summary"></a>Summary</h2>
    </div>

    
    
    
        

<div>
<ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>本文作者： </strong>Jack Zheng
  </li>
  <li class="post-copyright-link">
    <strong>本文链接：</strong>
    <a href="https://jack-zheng.github.io/hexo/2021/01/21/TIJ4-error-handing-with-exceptions/" title="TIJ4 通过异常处理错误 error handing with exceptions">https://jack-zheng.github.io/hexo/2021/01/21/TIJ4-error-handing-with-exceptions/</a>
  </li>
  <li class="post-copyright-license">
    <strong>版权声明： </strong>本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/null" rel="noopener" target="_blank"><i class="fa fa-fw fa-creative-commons"></i>BY-NC-SA</a> 许可协议。转载请注明出处！
  </li>
</ul>
</div>


      <footer class="post-footer">
          <div class="post-tags">
              <a href="/hexo/tags/exception/" rel="tag"># exception</a>
          </div>

        

          <div class="post-nav">
            <div class="post-nav-next post-nav-item">
                <a href="/hexo/2021/01/20/TIJ4-arrays/" rel="next" title="TIJ4 数组 Arrays">
                  <i class="fa fa-chevron-left"></i> TIJ4 数组 Arrays
                </a>
            </div>

            <span class="post-nav-divider"></span>

            <div class="post-nav-prev post-nav-item">
            </div>
          </div>
      </footer>
    
  </article>
  
  
  

  </div>


          </div>
          
    <div class="comments" id="gitalk-container"></div>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          文章目录
        </li>
        <li class="sidebar-nav-overview">
          站点概览
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#%E5%89%8D%E8%BF%B0"><span class="nav-number">1.</span> <span class="nav-text">前述</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Concepts"><span class="nav-number">2.</span> <span class="nav-text">Concepts</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Basic-exceptions"><span class="nav-number">3.</span> <span class="nav-text">Basic exceptions</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Exception-arguments"><span class="nav-number">4.</span> <span class="nav-text">Exception arguments</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Catching-an-exception"><span class="nav-number">5.</span> <span class="nav-text">Catching an exception</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#The-try-block"><span class="nav-number">5.1.</span> <span class="nav-text">The try block</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Exception-handlers"><span class="nav-number">5.2.</span> <span class="nav-text">Exception handlers</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Termination-vs-resumption-%E4%B8%AD%E6%96%AD%E8%BF%98%E6%98%AF%E7%BB%A7%E7%BB%AD%EF%BC%9F"><span class="nav-number">6.</span> <span class="nav-text">Termination vs. resumption (中断还是继续？)</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Creating-your-own-exceptions"><span class="nav-number">7.</span> <span class="nav-text">Creating your own exceptions</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Exercises"><span class="nav-number">8.</span> <span class="nav-text">Exercises</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Exceptions-and-logging"><span class="nav-number">9.</span> <span class="nav-text">Exceptions and logging</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#The-exception-specification"><span class="nav-number">10.</span> <span class="nav-text">The exception specification</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Catching-any-exception"><span class="nav-number">11.</span> <span class="nav-text">Catching any exception</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#The-stack-trace"><span class="nav-number">11.1.</span> <span class="nav-text">The stack trace</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Rethrowing-an-exception"><span class="nav-number">11.2.</span> <span class="nav-text">Rethrowing an exception</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Exception-chaining"><span class="nav-number">11.3.</span> <span class="nav-text">Exception chaining</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Standard-Java-exceptions"><span class="nav-number">12.</span> <span class="nav-text">Standard Java exceptions</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Special-case-RuntimeException"><span class="nav-number">12.1.</span> <span class="nav-text">Special case: RuntimeException</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Performing-cleanup-with-finally"><span class="nav-number">13.</span> <span class="nav-text">Performing cleanup with finally</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#What%E2%80%99s-finally-for"><span class="nav-number">13.1.</span> <span class="nav-text">What’s finally for?</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Using-finally-during-return"><span class="nav-number">13.2.</span> <span class="nav-text">Using finally during return</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Pitfall-the-lost-exception"><span class="nav-number">13.3.</span> <span class="nav-text">Pitfall: the lost exception</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Exception-restrictions"><span class="nav-number">14.</span> <span class="nav-text">Exception restrictions</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Constructors"><span class="nav-number">15.</span> <span class="nav-text">Constructors</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Exception-matching"><span class="nav-number">16.</span> <span class="nav-text">Exception matching</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Alternative-approaches"><span class="nav-number">17.</span> <span class="nav-text">Alternative approaches</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#History"><span class="nav-number">17.1.</span> <span class="nav-text">History</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Perspectives"><span class="nav-number">17.2.</span> <span class="nav-text">Perspectives</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Passing-exceptions-to-the-console"><span class="nav-number">17.3.</span> <span class="nav-text">Passing exceptions to the console</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Converting-checked-to-unchecked-exceptions"><span class="nav-number">17.4.</span> <span class="nav-text">Converting checked to unchecked exceptions</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Exception-guidelines"><span class="nav-number">18.</span> <span class="nav-text">Exception guidelines</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Summary"><span class="nav-number">19.</span> <span class="nav-text">Summary</span></a></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Jack Zheng"
      src="/hexo/images/jack.jpg">
  <p class="site-author-name" itemprop="name">Jack Zheng</p>
  <div class="site-description" itemprop="description">我是张江搬砖小码农, 知识搬运工</div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/hexo/archives/">
        
          <span class="site-state-item-count">105</span>
          <span class="site-state-item-name">日志</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/hexo/categories/">
          
        <span class="site-state-item-count">10</span>
        <span class="site-state-item-name">分类</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/hexo/tags/">
          
        <span class="site-state-item-count">94</span>
        <span class="site-state-item-name">标签</span></a>
      </div>
  </nav>
</div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2021</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Jack Zheng</span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-area-chart"></i>
    </span>
    <span title="站点总字数">394k</span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-coffee"></i>
    </span>
    <span title="站点阅读时长">5:58</span>
</div>
  <div class="powered-by">由 <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> 强力驱动 v5.0.0
  </div>
  <span class="post-meta-divider">|</span>
  <div class="theme-info">主题 – <a href="https://muse.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Muse</a> v7.5.0
  </div>

        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <span class="post-meta-item" id="busuanzi_container_site_uv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-user"></i>
      </span>
      <span class="site-uv" title="总访客量">
        <span id="busuanzi_value_site_uv"></span>
      </span>
    </span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item" id="busuanzi_container_site_pv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-eye"></i>
      </span>
      <span class="site-pv" title="总访问量">
        <span id="busuanzi_value_site_pv"></span>
      </span>
    </span>
</div>












        
      </div>
    </footer>
  </div>

  
  <script src="/hexo/lib/anime.min.js"></script>
  <script src="/hexo/lib/velocity/velocity.min.js"></script>
  <script src="/hexo/lib/velocity/velocity.ui.min.js"></script>

<script src="/hexo/js/utils.js"></script>

<script src="/hexo/js/motion.js"></script>


<script src="/hexo/js/schemes/muse.js"></script>


<script src="/hexo/js/next-boot.js"></script>




  




  
<script src="/hexo/js/local-search.js"></script>














  

  

<link rel="stylesheet" href="//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.css">

<script>
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js', () => {
    var gitalk = new Gitalk({
      clientID: 'd444c814810f7e73ee3a',
      clientSecret: '9b3c417a0c1076887b3cf583c7d7a27456b708ef',
      repo: 'hexo-comments',
      owner: 'jack-zheng',
      admin: ['jack-zheng'],
      id: '973149845d68b14113006943a2902f03',
        language: '',
      distractionFreeMode: 'true'
    });
    gitalk.render('gitalk-container');
  }, window.Gitalk);
</script>

</body>
</html>
